### PR TITLE
Preprocessor class implementation

### PR DESCRIPTION
**Pull request by dummyindex**
In this new version, we make `Preprocessor` class to allow users to freely explore different preprocessing recipes whose step parameters that can be configured inside `Preprocessor`. Existing recipes in preprocessor includes `monocle`, `pearson residuals`, `seurat` and `sctransform`. Users can replace each preprocessing step with their own implementation with ease. For instance, `Preprocessor`'s monocle pipeline contains `filter_cells_by_outliers`, `filter_genes_by_outliers`, `normalize_by_cells`, `select_genes` and other modular preprocess steps. Users can replace the implementation and default monocle parameters passed in to these functions by replacing or changing attributes of `Preprocessor`.

### CODE REVIEW

File: dynamo/external/sctransform.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
+# =================================================================
+# Original Code Repository Author: Lause, Berens & Kobak
+# Adapted to Dynamo by: dynamo authors
+# Created Date: 12/16/2021
+# Description: pearson residuals based method for preprocessing single cell expression data
+# Original Code Repository: https://github.com/atarashansky/SCTransformPy
```
Comment (Xiaojieqiu) [inline dynamo/external/sctransform.py:6]: sctransform is from seurat originally, right? cite seurat's paper / package, etc?

File: dynamo/external/sctransform.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+# =================================================================
+
+import statsmodels.nonparametric.kernel_regression
+from KDEpy import FFTKDE
```
Comment (Xiaojieqiu) [inline dynamo/external/sctransform.py:10]: make sure KDEpy will be added to requirements.txt, etc.

File: dynamo/preprocessing/Preprocessor.py [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+        self.normalize_selected_genes_kwargs = {"layers": ["X"], "copy": False}
+
+        self.pca_kwargs = {"pca_key": "X_pca", "n_pca_components": 50}
+        self.use_log1p = False
+
+    def preprocess_adata_monocle_pearson_residuals(
+        self, adata: AnnData, tkey: Optional[str] = None, experiment_type: Optional[str] = None
+    ):
+        """
+        A combined pipeline of monocle and pearson_residuals.
```
Comment (Xiaojieqiu) [inline dynamo/preprocessing/Preprocessor.py:469]: pearson residual is used for adata.X but monocle for spliced/unspliced layers, right?

**Review by Xiaojieqiu — COMMENTED** [review]
I have reviewed the code. Can you make sure we can do something like this, instead of using the preprocessing class? 
```

    dyn.pp.filter_genes(adata, min_cells=min_cells)
    dyn.pp.normalize_total(adata)
    dyn.pp.log1p(adata)
    dyn.pp.pca(adata, n_comps=30)

```

File: dynamo/configuration.py [hunk 4]
```diff
@@ -1,9 +1,138 @@
...
+
+    def set_layer_data(adata: AnnData, layer: str, vals, var_indices: np.array = None):
+        if var_indices is None:
+            var_indices = slice(None)
+        if layer == DynamoAdataKeyManager.X_LAYER:
+            adata.X[:, var_indices] = vals
+        elif layer in adata.layers:
+            adata.layers[layer][:, var_indices] = vals
+        else:
+            # layer does not exist in adata
```
[the remark this answers is not in the document]
Comment (dummyindex) [reply dynamo/configuration.py:82]: sometimes only partial results are available. E.g. `use_for_pca` gene only results. In these cases if adata reserve all genes (keep all genes) as what is default in recipe_monocle, we need to set partial adata layer.

File: dynamo/external/sctransform.py (hunk 1 above)
Comment (dummyindex) [reply dynamo/external/sctransform.py:6]: Cited in preprocessor's preprocess_adata_sctransform method

File: dynamo/external/sctransform.py (hunk 2 above)
Comment (dummyindex) [reply dynamo/external/sctransform.py:10]: added to requirements.txt

File: dynamo/preprocessing/Preprocessor.py (hunk 3 above)
Comment (dummyindex) [reply dynamo/preprocessing/Preprocessor.py:469]: `preprocess_adata_monocle_pearson_residuals` does what you describe.  `pearson_residuals` recipe normalizes all raw data layers (new/total, spliced/unspliced, X). Users can specify what to normalize by speicfy `self.normalize_selected_genes_kwargs`

File: dynamo/external/sctransform.py (hunk 1 above)
Comment (dummyindex) [reply dynamo/external/sctransform.py:6]: added to this file as well in new commits.

File: dynamo/preprocessing/Preprocessor.py (hunk 3 above)
Comment (Xiaojieqiu) [reply dynamo/preprocessing/Preprocessor.py:506]: ok, got it. this sounds good to me.

File: dynamo/configuration.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Xiaojieqiu) [reply dynamo/configuration.py:82]: sounds good to me.

**Pull request merged**